factory_reset/reset.proto

```diff
@@ -14,28 +14,22 @@ service FactoryReset {
   // shipped from factory. State includes storage, configuration, logs,
   // certificates and licenses.
   //
-  // It is optional for the Target to support rolling back the OS to the same
-  // version shipped from factory. If not supported, the running OS should be
-  // used.
+  // Optionally allows rolling back the OS to the same version shipped from
+  // factory.
   //
-  // It is optional for the Target to zero fill permanent storage where state
+  // Optionally allows for the Target to zero-fill permanent storage where state
   // data is stored.
   rpc Start(StartRequest) returns (StartResponse);
 }
 
 message StartRequest {
   // Instructs the Target to rollback the OS to the same version as it shipped
-  // from factory.
+  // from factory. Target returns grpc status code UNIMPLEMENTED if set but not
```

Comment:
Did you consider INVALID_ARGUMENT so a caller could easily differentiate from the whole RPC not being implemented? The gRPC libraries can return UNIMPLEMENTED for certain conditions but will never return INVALID_ARGUMENT according to https://github.com/grpc/grpc/blob/master/doc/statuscodes.md

Reply:
Thanks! Updated.

I wonder if it is not less ambiguous and more precise to define error codes in the service directly. In practice a Client will perform an RPC call with two of the not implemented bits set. It will receive a gRPC error with INVALID_ARGUMENT and some arbitrary string. It will need to then parse that string to figure out which one of the optional arguments is not supported and try again with the given option unselected. A Client will then be at the mercy of the implementation done by the Target. This problem would not occur If this is done with explicit error codes defined in the service, example:
https://github.com/openconfig/gnoi/blob/master/os/os.proto#L213

In general I am in favour of precise services that reduce guessing at the client implementation.

Reply:
I am in full agreement on having explicit error conditions.  My original concern was with the semantic of performing the reset even if one of the requested options was unsupported. I think it would be good to explicitly return an error at the gRPC layer and if there are multiple differentiable invalid arguments that might need to be handled by the client in different ways also include the response message as before.

Reply:
so it turns out that if gRPC returns anything than !OK, gRPC does not guarantee the delivery of the RPC response message.
https://github.com/grpc/grpc/issues/16786#issuecomment-437878961

This is a generalization that will depend on the !OK type and the implementation, however it feels significant enougth to introduce ambiguity. And to avoid it, in order to use gRPC errors, only code and status should be used to provide extra information about the error. This brings ofcouse an assured loss of programatic detail about the nature of the error.

With this I am inclined more and more to not throwing a gRPC error and keeping the Error service specific. Made a new commit to reflect this. What are your thoughts?

Reply:
I have updated the spec to return gRPC Status error code if any of the optional functionalities are not supported/implemented.